docker-images/build.sh

```diff
@@ -63,20 +65,67 @@ function build {
     # Images depending on Kafka version (possibly indirectly thru FROM)
     for kafka_version in "${!version_checksums[@]}"
     do
-        sha=${version_checksums[$kafka_version]}
+        echo "Creating container image for Kafka $kafka_version"
+
+        expected_sha=${version_checksums[$kafka_version]}
         lib_directory=${version_libs[$kafka_version]}
+
+        binary_file_dir="./$kafka_image/tmp"
+
+        # If there is a file specified for this version of Kafka use that instead of the specified URL
+        if [ ${version_binary_files[$kafka_version]} ]
+        then 
+
+            custom_binary_file_path=${version_binary_files[$kafka_version]}
+            binary_file_name=$(basename "$custom_binary_file_path")
+            binary_file_path="$binary_file_dir/$binary_file_name"
+            echo "Kafka binary file field was specified, using $custom_binary_file_path instead of URL"
+
+            if [ -f "$binary_file_path" ]
+            then
+                echo "$binary_file_name already present in build directory"
```

Comment:
Should we do a checksum on the existing file here? If so should we be comparing it to the one specified in the path/url or to the sha from from the versions file?

Reply:
The sha in the yaml file is what we _trust_, so that should be what we compare with.

Reply:
Ok. If the sha does not match the already present file should we exit, or just delete that file and copy in the one from the url/filepath?

Reply:
* We must check, somewhere, that what ends up in the images (whether downloaded or copied from a file) matches the sha from the yaml. That can be done either just prior to the docker build, or within the docker build. 

* We may also use the sha as a way to avoid downloading a file which we already have locally. Since copying a copy from one place to another is approximately free there's not much benefit to be had from avoiding the copy, but there's no harm either.

So I would check whether any file already in the build dir matches the sha and if so the we're done. Otherwise we need to download/copy and then check again.

Reply:
The script currently always checksums the file, whether it was already present or not, in the build dir just before the build, so I think this is good to go?

Reply:
Actually it will not download/copy if the checksum fails. This will require adding some functions to make the code cleaner.

Reply:
I don't think you need a function. Can't you just do something like

```
if [ $expected_sha != $(sha1sum $file) ] 
then 
  # download or copy
fi
```

and restore the sha check in the `Dockerfile`. That would suffice, right?

Reply:
Ok so the existing files are now checksummed. This involved using `shopt -s nocasematch` and using `[[` instead of `[` as the output of `sha512sum` is all lowercase and our checksums are uppercase. This is certainly a journey of discovery.